### PR TITLE
fix: touching of unions

### LINKED ISSUE #2905: concatenation + combinations leaves placeholder arrays in data after form rehydration?
### Version of Awkward Array

2.5.1 / coffea 2023.12.0

### Description and code to reproduce

The following code (you will need to pip install coffea):
```python3
from coffea.nanoevents import NanoEventsFactory, NanoAODSchema
import awkward as ak

if __name__ == "__main__":
    NanoAODSchema.error_missing_event_ids = False

    events = NanoEventsFactory.from_root(
        {"tests/samples/nano_dy.root": {"object_path": "Events", "steps": [[0,40]]}},
        metadata={"dataset": "nano_dy"},
        schemaclass=NanoAODSchema,
    ).events()

    events["leptons"] = ak.concatenate(
        [events.Electron, events.Muon],
        axis=1,
    )

    pair = ak.argcombinations(events.leptons, 2, fields=["l1", "l2"])

    print((events.leptons[pair.l1] + events.leptons[pair.l2]).pt.compute())
```
fails with:
```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
File ~/miniforge3/envs/coffea_2023_check/lib/python3.10/site-packages/awkward/highlevel.py:1511, in Array.__array_ufunc__(self, ufunc, method, *inputs, **kwargs)
   1510 with ak._errors.OperationErrorContext(name, inputs, kwargs):
-> 1511     return ak._connect.numpy.array_ufunc(ufunc, method, inputs, kwargs)

File ~/miniforge3/envs/coffea_2023_check/lib/python3.10/site-packages/awkward/_connect/numpy.py:458, in array_ufunc(ufunc, method, inputs, kwargs)
    456     return None
--> 458 out = ak._broadcasting.broadcast_and_apply(
    459     inputs, action, allow_records=False, function_name=ufunc.__name__
    460 )
    462 if len(out) == 1:

File ~/miniforge3/envs/coffea_2023_check/lib/python3.10/site-packages/awkward/_broadcasting.py:1027, in broadcast_and_apply(inputs, action, depth_context, lateral_context, allow_records, left_broadcast, right_broadcast, numpy_to_regular, regular_to_jagged, function_name, broadcast_parameters_rule)
   1026 isscalar = []
-> 1027 out = apply_step(
   1028     backend,
   1029     broadcast_pack(inputs, isscalar),
   1030     action,
   1031     0,
   1032     depth_context,
   1033     lateral_context,
   1034     {
   1035         "allow_records": allow_records,
   1036         "left_broadcast": left_broadcast,
   1037         "right_broadcast": right_broadcast,
   1038         "numpy_to_regular": numpy_to_regular,
   1039         "regular_to_jagged": regular_to_jagged,
   1040         "function_name": function_name,
   1041         "broadcast_parameters_rule": broadcast_parameters_rule,
   1042     },
   1043 )
   1044 assert isinstance(out, tuple)

File ~/miniforge3/envs/coffea_2023_check/lib/python3.10/site-packages/awkward/_broadcasting.py:1005, in apply_step(backend, inputs, action, depth, depth_context, lateral_context, options)
   1004 elif result is None:
-> 1005     return continuation()
   1006 else:

File ~/miniforge3/envs/coffea_2023_check/lib/python3.10/site-packages/awkward/_broadcasting.py:974, in apply_step.<locals>.continuation()
    973 elif any(x.is_list and not is_string_like(x) for x in contents):
--> 974     return broadcast_any_list()
    976 # Any RecordArrays?

File ~/miniforge3/envs/coffea_2023_check/lib/python3.10/site-packages/awkward/_broadcasting.py:630, in apply_step.<locals>.broadcast_any_list()
    628         nextparameters.append(NO_PARAMETERS)
--> 630 outcontent = apply_step(
    631     backend,
    632     nextinputs,
    633     action,
    634     depth + 1,
    635     copy.copy(depth_context),
    636     lateral_context,
    637     options,
    638 )
    639 assert isinstance(outcontent, tuple)

File ~/miniforge3/envs/coffea_2023_check/lib/python3.10/site-packages/awkward/_broadcasting.py:1005, in apply_step(backend, inputs, action, depth, depth_context, lateral_context, options)
   1004 elif result is None:
-> 1005     return continuation()
   1006 else:

File ~/miniforge3/envs/coffea_2023_check/lib/python3.10/site-packages/awkward/_broadcasting.py:974, in apply_step.<locals>.continuation()
    973 elif any(x.is_list and not is_string_like(x) for x in contents):
--> 974     return broadcast_any_list()
    976 # Any RecordArrays?

File ~/miniforge3/envs/coffea_2023_check/lib/python3.10/site-packages/awkward/_broadcasting.py:694, in apply_step.<locals>.broadcast_any_list()
    692         nextparameters.append(NO_PARAMETERS)
--> 694 outcontent = apply_step(
    695     backend,
    696     nextinputs,
    697     action,
    698     depth + 1,
    699     copy.copy(depth_context),
    700     lateral_context,
    701     options,
    702 )
    703 assert isinstance(outcontent, tuple)

File ~/miniforge3/envs/coffea_2023_check/lib/python3.10/site-packages/awkward/_broadcasting.py:1005, in apply_step(backend, inputs, action, depth, depth_context, lateral_context, options)
   1004 elif result is None:
-> 1005     return continuation()
   1006 else:

File ~/miniforge3/envs/coffea_2023_check/lib/python3.10/site-packages/awkward/_broadcasting.py:966, in apply_step.<locals>.continuation()
    965 elif any(x.is_union for x in contents):
--> 966     return broadcast_any_union()
    968 # Any option-types?

File ~/miniforge3/envs/coffea_2023_check/lib/python3.10/site-packages/awkward/_broadcasting.py:867, in apply_step.<locals>.broadcast_any_union()
    866 if isinstance(x, UnionArray):
--> 867     nextinputs.append(x[mask].project(combo[i]))
    868     i += 1

File ~/miniforge3/envs/coffea_2023_check/lib/python3.10/site-packages/awkward/contents/unionarray.py:711, in UnionArray.project(self, index)
    708 nextcarry = ak.index.Index64(
    709     tmpcarry.data[: lenout[0]], nplike=self._backend.index_nplike
    710 )
--> 711 return self._contents[index]._carry(nextcarry, False)

File ~/miniforge3/envs/coffea_2023_check/lib/python3.10/site-packages/awkward/contents/recordarray.py:532, in RecordArray._carry(self, carry, allow_lazy)
    531 else:
--> 532     contents = [
    533         self.content(i)._carry(carry, allow_lazy)
    534         for i in range(len(self._contents))
    535     ]
    537     # if issubclass(carry.dtype.type, np.integer):
    538     #     length = carry.length
    539     # else:
    540     #     length = self.length

File ~/miniforge3/envs/coffea_2023_check/lib/python3.10/site-packages/awkward/contents/recordarray.py:533, in <listcomp>(.0)
    531 else:
    532     contents = [
--> 533         self.content(i)._carry(carry, allow_lazy)
    534         for i in range(len(self._contents))
    535     ]
    537     # if issubclass(carry.dtype.type, np.integer):
    538     #     length = carry.length
    539     # else:
    540     #     length = self.length

File ~/miniforge3/envs/coffea_2023_check/lib/python3.10/site-packages/awkward/contents/numpyarray.py:347, in NumpyArray._carry(self, carry, allow_lazy)
    346 try:
--> 347     nextdata = self._data[carry.data]
    348 except IndexError as err:

File ~/miniforge3/envs/coffea_2023_check/lib/python3.10/site-packages/awkward/_nplikes/placeholder.py:97, in PlaceholderArray.__getitem__(self, index)
     96 else:
---> 97     raise TypeError(
     98         f"{type(self).__name__} supports only trivial slices, not {type(index).__name__}"
     99     )

TypeError: PlaceholderArray supports only trivial slices, not ndarray

The above exception was the direct cause of the following exception:

TypeError                                 Traceback (most recent call last)
Cell In[2], line 40
     31 mt = np.sqrt(2 * l3.pt * events.MET.pt * (1 - np.cos(events.MET.delta_phi(l3))))
     32 q8_hist = (
     33     hda.Hist.new.Reg(
     34         100, 0, 200, name="mt", label="$\ell$-MET transverse mass [GeV]"
   (...)
     37     .fill(mt)
     38 )
---> 40 q8_hist.compute().plot1d()
     41 dak.necessary_columns(q8_hist)

File ~/miniforge3/envs/coffea_2023_check/lib/python3.10/site-packages/dask/base.py:342, in DaskMethodsMixin.compute(self, **kwargs)
    318 def compute(self, **kwargs):
    319     """Compute this dask collection
    320 
    321     This turns a lazy Dask collection into its in-memory equivalent.
   (...)
    340     dask.compute
    341     """
--> 342     (result,) = compute(self, traverse=False, **kwargs)
    343     return result

File ~/miniforge3/envs/coffea_2023_check/lib/python3.10/site-packages/dask/base.py:628, in compute(traverse, optimize_graph, scheduler, get, *args, **kwargs)
    625     postcomputes.append(x.__dask_postcompute__())
    627 with shorten_traceback():
--> 628     results = schedule(dsk, keys, **kwargs)
    630 return repack([f(r, *a) for r, (f, a) in zip(results, postcomputes)])

File ~/miniforge3/envs/coffea_2023_check/lib/python3.10/site-packages/dask_awkward/lib/core.py:1883, in ArgsKwargsPackedFunction.__call__(self, *args_deps_expanded)
   1881     len_args += n_args
   1882 kwargs = self.kwarg_repacker(args_deps_expanded[len_args:])[0]
-> 1883 return self.fn(*args, **kwargs)

File ~/miniforge3/envs/coffea_2023_check/lib/python3.10/site-packages/awkward/highlevel.py:1510, in Array.__array_ufunc__(self, ufunc, method, *inputs, **kwargs)
   1445 """
   1446 Intercepts attempts to pass this Array to a NumPy
   1447 [universal functions](https://docs.scipy.org/doc/numpy/reference/ufuncs.html)
   (...)
   1507 See also #__array_function__.
   1508 """
   1509 name = f"{type(ufunc).__module__}.{ufunc.__name__}.{method!s}"
-> 1510 with ak._errors.OperationErrorContext(name, inputs, kwargs):
   1511     return ak._connect.numpy.array_ufunc(ufunc, method, inputs, kwargs)

File ~/miniforge3/envs/coffea_2023_check/lib/python3.10/site-packages/awkward/_errors.py:85, in ErrorContext.__exit__(self, exception_type, exception_value, traceback)
     78 try:
     79     # Handle caught exception
     80     if (
     81         exception_type is not None
     82         and issubclass(exception_type, Exception)
     83         and self.primary() is self
     84     ):
---> 85         self.handle_exception(exception_type, exception_value)
     86 finally:
     87     # Step out of the way so that another ErrorContext can become primary.
     88     if self.primary() is self:

File ~/miniforge3/envs/coffea_2023_check/lib/python3.10/site-packages/awkward/_errors.py:95, in ErrorContext.handle_exception(self, cls, exception)
     93     self.decorate_exception(cls, exception)
     94 else:
---> 95     raise self.decorate_exception(cls, exception)

TypeError: PlaceholderArray supports only trivial slices, not ndarray

This error occurred while calling

    numpy.add.__call__(
        repr-raised-TypeError
        repr-raised-TypeError
    )
```

So we're hitting a placeholder array somewhere in the calculation, which shouldn't be there, it seems?

It appears to be something with the concatenate operation:
```python3
from coffea.nanoevents import NanoEventsFactory, NanoAODSchema
import awkward as ak

if __name__ == "__main__":
    NanoAODSchema.error_missing_event_ids = False

    events = NanoEventsFactory.from_root(
        {"tests/samples/nano_dy.root": {"object_path": "Events", "steps": [[0,40]]}},
        metadata={"dataset": "nano_dy"},
        schemaclass=NanoAODSchema,
    ).events()

    events["leptons"] = events.Electron

    pair = ak.argcombinations(events.leptons, 2, fields=["l1", "l2"])

    print((events.leptons[pair.l1] + events.leptons[pair.l2]).pt.compute())
```
Executes successfully.

@agoose77 @jpivarski reasonably high priority here since this sort of pattern is common in realistic analysis scenarios.

### PR DESCRIPTION
**Pull request by agoose77**
Broadcasting of unions should also touch the contents! Fixes #2905

### CODE REVIEW

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -61,7 +61,6 @@ jobs:
 
     env:
       PIP_ONLY_BINARY: numpy,pandas,pyarrow,numexpr,numexpr
-      PIP_PRE: "1"
```
Comment (agoose77) [inline .github/workflows/test.yml]: We don't want any pre-releases by default.

File: src/awkward/_broadcasting.py [hunk 2]
```diff
@@ -773,121 +773,74 @@ def broadcast_any_union():
...
-                outcontents.append(
-                    apply_step(
-                        backend,
-                        nextinputs,
-                        action,
-                        depth,
-                        copy.copy(depth_context),
-                        lateral_context,
-                        options,
+        union_tags, union_num_contents, length = [], [], unknown_length
```
Comment (agoose77) [inline src/awkward/_broadcasting.py:776]: All of the subsequent code is slightly refactored to drop out the branch!

File: src/awkward/contents/bytemaskedarray.py [hunk 3]
```diff
@@ -390,7 +390,7 @@ def _getitem_at(self, where: IndexType):
         else:
             return None
 
-    def _getitem_range(self, start: SupportsIndex, stop: IndexType) -> Content:
+    def _getitem_range(self, start: IndexType, stop: IndexType) -> Content:
```
Comment (agoose77) [inline src/awkward/contents/bytemaskedarray.py:393]: Fix type signatures (mypy)

File: src/awkward/index.py [hunk 4]
```diff
@@ -214,6 +215,9 @@ def form(self) -> str:
         return _dtype_to_form[self._data.dtype]
 
     def __getitem__(self, where):
+        if isinstance(where, slice):
+            where = normalize_slice(where, nplike=self.nplike)
```
Comment (agoose77) [inline src/awkward/index.py:219]: We need to normalize the slice so that `unknown_length` becomes an unknown scalar.

**Review by jpivarski — APPROVED** [review]
That's right: since an element of a union can be any of its contents, and it doesn't get unpacked by the user explicitly specifying a field name, touching a union must touch all of its contents.

It looks like this PR also has the CI improvements. Maybe that's because it has been merged into this but not `main` yet.

**Pull request merged**